### PR TITLE
Feature: DKG State

### PR DESCRIPTION
**Pull request by jordanschalm**
Extends the DKGState to store end states. 

* https://github.com/onflow/flow-go/pull/1720
* https://github.com/onflow/flow-go/pull/1721
* https://github.com/onflow/flow-go/pull/1732

### CODE REVIEW

**Review by zhangchiqing — APPROVED** [review]

File: engine/consensus/dkg/reactor_engine.go [hunk 1]
```diff
@@ -161,34 +229,90 @@ func (e *ReactorEngine) EpochSetupPhaseStarted(currentEpochCounter uint64, first
...
 
-	localPubKey := dkgPrivInfo.PublicKey()
-
+	// we computed a local beacon key but it is inconsistent with our canonical
+	// public key - therefore it is unsafe for use
 	if !nextDKGPubKey.Equals(localPubKey) {
-		e.log.Warn().Msg("checking DKG key consistency: locally computed dkg public key does not match dkg public key for next epoch")
+		log.Warn().
+			Hex("computed_beacon_pub_key", localPubKey.Encode()).
+			Hex("canonical_beacon_pub_key", nextDKGPubKey.Encode()).
```
Comment (tarakby) [inline engine/consensus/dkg/reactor_engine.go:302]: We could use the `String()` function of public keys

File: engine/consensus/dkg/reactor_engine.go [hunk 2]
```diff
@@ -161,34 +229,90 @@ func (e *ReactorEngine) EpochSetupPhaseStarted(currentEpochCounter uint64, first
...
+			log.Fatal().Err(err).Msg("failed to set dkg end state")
+		}
+		return
 	}
+
+	err = e.dkgState.SetDKGEndState(nextEpochCounter, flow.DKGEndStateSuccess)
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("failed to set dkg")
+	}
+	log.Info().Msgf("successfully ended DKG, my beacon pub key for epoch %d is %x", nextEpochCounter, localPubKey.Encode())
```
Comment (tarakby) [inline engine/consensus/dkg/reactor_engine.go:315]: I guess this should work
```suggestion
	log.Info().Msgf("successfully ended DKG, my beacon pub key for epoch %d is %s", nextEpochCounter, localPubKey)
```

**Review by tarakby — APPROVED** [review]
This looks clean 👌🏼  nice work!

**Comment by jordanschalm** [comment]
bors merge

**Pull request merged**